### ISSUE TITLE
Microsoft Teams crashes when sharing screen under Wayland

### ISSUE THREAD

**Issue by MightyCreak** [electron/electron#23063]
### Preflight Checklist

* [x] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
* [x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
* [x] I have searched the issue tracker for an issue that matches the one I want to file, without success.

### Issue Details

* **Electron Version:**
  * N/A (I'm sorry, I can't get to know which version of Electron is used by Microsoft Teams)
* **Operating System:**
  * Fedora 31

### Expected Behavior
Be able to share my screen in the Teams application.

### Actual Behavior
Sharing the screen when on Wayland is actually crashing the Teams application.

### To Reproduce
* Start a session under GNOME (Wayland)
* Install flatpak package of Teams: `flatpak install com.microsoft.Teams`
* Run Teams
* Start a video call
* Share your screen
* Teams crashes and closes

### Additional Information
* Workaround: It works when using Xorg.
* Teams Flatpak package repo: https://github.com/flathub/com.microsoft.Teams/
* I first posted an [issue on Teams flatpak repo](https://github.com/flathub/com.microsoft.Teams/issues/23#issuecomment-611241691) and, considering the answers, I decided to open a new bug here

**Comment by VerteDinde** [comment]
Hey @MightyCreak! We appreciate you taking the time to report this, but this unfortunately is a Microsoft Teams-specific bug, not an Electron-specific bug, and so is not in our control to fix unless we can get more information from the developers.

You can report the bug to Microsoft Teams directly here: https://docs.microsoft.com/en-us/microsoftteams/platform/feedback

**Comment by MightyCreak** [comment]
Hi @VerteDinde from [the comment](https://github.com/flathub/com.microsoft.Teams/issues/23#issuecomment-611241691) it seems that screen sharing doesn't work with any Electron app on Wayland. I haven't tested another app right that would have this feature, but if this is true, it seems that it would be an Electron bug. Is screen sharing working for Slack when on Wayland for instance?

**Comment by danshick** [comment]
Any chance of adding the 'rtc_use_pipewire=true' build flag to support Wayland screen sharing?

**Comment by titaniumbones** [comment]
@danshick is that a question for electron or for MS? I'm trying to figure out where I should follow along to track this issue.

**Comment by danshick** [comment]
It's both.

Electron needs a fix and MS needs to build a version of Teams which includes that fix.

As I understand it, the people working on a wayland compatible electron (called electron-ozone) have already added this fix, but electron-ozone itself is not that stable.

Electron doesn't need to wait for ozone to add this fix though. This flag works fine from xwayland, as evidenced by Chromium. Even still, it has to be added, tested, make it into a release, adopted by MS in a build of Teams, and the logic about when to display the screenshare button needs to be updated.

For now, use browser-based Teams in Chromium that supports PipeWire, and set a different user agent so the screenshare button shows up.

**Comment by titaniumbones** [comment]
THanks @danshick . Sounds pretty awkward, will maybe just not use the tool :-(

**Comment by a-pavlenko** [comment]
Screen sharing works in [unofficial Teams app](https://github.com/IsmaelMartinez/teams-for-linux), which also uses electron.

**Comment by titaniumbones** [comment]
@a-pavlenko screen sharing works on wayland? I think the OP and I have both described a bug with the same app you're referring to, so this would be a great new development!

**Comment by a-pavlenko** [comment]
@titaniumbones it seems OP mentioned Teams flatpak app. 

Are you sure you are using the unofficial client from the link above? Because that's the only app I got working on Wayland. However, I can share only a window, not the entire screen.

**Comment by danshick** [comment]
I doubt very much that any Electron app is capable of full screen screen sharing on Wayland yet unless you are using electron-ozone or building electron yourself. Electron sharing will work with XWayland windows as well as other Chrome tabs or Electron apps, and this was always the case.

**Comment by MightyCreak** [comment]
As @a-pavlenko said, there are two flatpak packages for Teams:
1. https://flathub.org/apps/details/com.microsoft.Teams (the "most" official one, takes the official Debian package, updated last March)
2. https://flathub.org/apps/details/com.github.IsmaelMartinez.teams_for_linux (wrap the website in an Electron app, updated last October)

Only the first one is the one used for this issue since it is using the official Debian package made by Microsoft.

**Comment by danshick** [comment]
@MightyCreak Yes, this particular issue relates to the official Teams app only, but neither version will correctly support screen sharing as Electron doesn't build with RTC_USE_PIPEWIRE=true by default. This is true of every electron app using upstream electron at the moment.

**Comment by btd1337** [comment]
After `Electron 8.2.2` version the `rtc_use_pipewire=true` build flag to support Wayland screen sharing.

Reference: https://github.com/electron/electron/issues/10915

**Comment by danshick** [comment]
This is only true in the experimental ozone branch of electron I believe. That issue is still open. What version of electron does Teams build against?

**Comment by titaniumbones** [comment]
As an end user I just want to be sure I understand the status here: It's not possible to build/install electron-ozone and somehow require teams to use that version? From what I can see the deb package  source includes only binaries but wanted to be sure. Thanks!

**Comment by danshick** [comment]
That seems like a lot more work than just using the browser based version which can screenshare (assuming you have a browser built with RTC_USE_PIPEWIRE=true).

**Comment by SpineEyE** [comment]
Thr browser version only shows one video in conference calls, which is simply not enough. The electron app on Linux also shows only 4 and not 9 or 49, but it's still much better than 1.

**Comment by danshick** [comment]
I'd bug MS about that. Maybe try faking your user agent? I find it unacceptable that software would behave differently in different browsers and OSes.

Regardless, I very much doubt it is worth trying to get the packaged Teams app to work with a different electron binary. Even if it were open source, it could be non-trivial to build against a different version of Electron.

For instance, I know the electron-ozone folks struggle to run VSCode against their builds.

Most importantly, this issue, as it relates to a change that the **electron project** can make to better support screensharing in Teams as well as all other electron apps would be to add the appropriate build flag for pipewire in upstream electron.

**Comment by klaasjanelzinga** [comment]
Wrt teams, I posted a uservoice on their site. Please vote :) 

https://microsoftteams.uservoice.com/forums/555103-public/suggestions/41524504-screen-sharing-on-wayland

**Comment by danshick** [comment]
> Wrt teams, I posted a uservoice on their site for this issue. Please vote :) 

There's nothing MS can do about this until a fix lands in Electron. I suppose they could contribute one, but it wouldn't necessarily be accepted upstream.

**Comment by Ella-0** [comment]
I have the exact same issue when using Ozone with Discord.

**Comment by jczaja** [comment]
I had the same problem (Fedora 32) , once I updated (today) teams to vesion 1.3.00.30857-1.x86_64  problem disappeared.

**Comment by bkhl** [comment]
> I had the same problem (Fedora 32) , once I updated (today) teams to vesion 1.3.00.30857-1.x86_64 problem disappeared.

I tried this as well today. I can share presentations without crashes, but I'm not given the option to share my screen.

**Comment by klingtnet** [comment]
> but I'm not given the option to share my screen.

I can second this for Arch Linux running Gnome shell 3.38 under Wayland.
My workaround is to use teams in chromium, which works pretty fine, event with screen sharing.  In fact, I do not see any advantage of using the electron app at the moment.

**Comment by paretje** [comment]
I use a chromium derivative, but I see two reasons to use the electron app: 
1. I get more then one video stream. This makes the webapp unusable to do a call with multiple users involved, especially due to the delay when switching between video streams. Well, except when you need to share your screen, then you need to switch quickly ...
2. I always get notifications. In my browser I don't seem to get notifications for the current chat, so it frequently happens that someone is speaking to me without me knowing. That might be an issue with my browser (qutebrowser), I didn't really debug it yet.

I would prefer to use my browser, but as the Linux electron app is still lagging behind, being limited to only 4 video streams, I'm not holding my breath for when decent support would land in the webapp ...

**Comment by zcbenz** [comment]
The build flag is enabled in Electron but the feature itself is disabled by default, can you try if passing `--enable-features=WebRTCPipeWireCapturer` in command line solves the problem?

**Comment by danshick** [comment]
> --enable-features=WebRTCPipeWireCapturer

I'm still getting pre-build flag behavior as of v10.1.5 and I don't see the flag in the code currently (at least not where I'd have thought it should go). Can you let me know if there's a specific version that I should be using?

**Comment by danshick** [comment]
> The build flag is enabled in Electron but the feature itself is disabled by default, can you try if passing `--enable-features=WebRTCPipeWireCapturer` in command line solves the problem?

I'm still getting pre-build flag behavior as of v10.1.5 and I don't see the flag in the code currently (at least not where I'd have thought it should go). Can you let me know if there's a specific version that I should be using?

**Comment by zcbenz** [comment]
The `rtc_use_pipewire=true` build flag is enabled by default in recent releases of Chromium, so you don't find it in Electron's code. I don't know which Electron version is used by Teams, and whether they explicitly disabled the build flag.

Anyway to enable PipeWire in Teams you need to ask Teams team to:
1. Make sure `rtc_use_pipewire` build flag is turned on.
2. Explicitly enable the `WebRTCPipeWireCapturer` feature in their codebase.

But still I don't know if this would solve the crash, if you want to have maintainers of Electron to be able to look into this issue, you have to work out a minimal example to reproduce the crash in Electron, without involving Teams. Otherwise you have to count on Teams developers to solve it for you.

**Comment by danshick** [comment]
> The `rtc_use_pipewire=true` build flag is enabled by default in recent releases of Chromium, so you don't find it in Electron's code. I don't know which Electron version is used by Teams, and whether they explicitly disabled the build flag.

A stabilizing commit for pipewire was recently reverted and abandoned (seeking to reduce the number of system dialogs, which was the reason for keeping the feature behind a flag).

Do you have a reference for the build flag being enabled by default currently? I'd like to add this to the compatibility wiki page we maintain for xdg-desktop-portal-wlr.

**Comment by danshick** [comment]
> But still I don't know if this would solve the crash, if you want to have maintainers of Electron to be able to look into this issue, you have to work out a minimal example to reproduce the crash in Electron, without involving Teams. Otherwise you have to count on Teams developers to solve it for you.

This issue really shouldn't be Teams specific, and should cover pipewire compatibility for all Electron apps. I'm experimenting with a minimal reproduction and getting a backtrace for the crashes I've experienced. Once that's shaped up, I'll create a new issue.

**Comment by danshick** [comment]
> The `rtc_use_pipewire=true` build flag is enabled by default in recent releases of Chromium, so you don't find it in Electron's code. I don't know which Electron version is used by Teams, and whether they explicitly disabled the build flag.

@zcbenz  I was just able to get Pipewire screensharing working in Electron, but it required that I build Electron with the `rtc_use_pipewire=true` flag explicitly, and it also required the `--enable-features=WebRTCPipeWireCapturer` argument. I am building the AUR package for [electron-ozone](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=electron-ozone), which includes this flag.

**Comment by zcbenz** [comment]
@danshick The `rtc_use_pipewire` is defined in https://source.chromium.org/chromium/chromium/src/+/master:third_party/webrtc/webrtc.gni;l=115?q=rtc_use_pipewire&sq=&ss=chromium
which is `true` in the official builds of Electron. Many third party builds set `use_sysroot=false` so `rtc_use_pipewire` might be `false` there.

**Comment by danshick** [comment]
> @danshick The `rtc_use_pipewire` is defined in https://source.chromium.org/chromium/chromium/src/+/master:third_party/webrtc/webrtc.gni;l=115?q=rtc_use_pipewire&sq=&ss=chromium
> which is `true` in the official builds of Electron. Many third party builds set `use_sysroot=false` so `rtc_use_pipewire` might be `false` there.

Thanks, that gives me something to check.

**Comment by marc-bouvier** [comment]
Could there be a relation with why Ubuntu 21.04 default to Wayland will fallback to X.org for nvidia users because of bad wayland nvidia support?
https://www.omgubuntu.co.uk/2021/01/ubuntu-21-04-will-use-wayland-by-default

**Comment by bkhl** [comment]
Don't think so, I get the same thing without any Nvidia hardware in sight.

**Comment by mehdisadeghi** [comment]
Electron 12 was release a few hours ago: https://github.com/electron/electron/releases/tag/v12.0.0

**Comment by electron-triage** [comment]
The Electron version reported on this issue is no longer supported. See our [supported versions documentation](https://www.electronjs.org/docs/tutorial/support#supported-versions).

**If you're still experiencing this issue on a supported version, please open a new issue with an updated repro - a [Fiddle](https://www.electronjs.org/fiddle) is very appreciated.**

Electron has a large issues backlog. To help our team prioritize, we're closing older issues and asking for new issues with updated repro steps if it affects a supported version. This helps sort what issues are still relevant and helps us fix them more quickly.

Thanks for your patience and understanding!

**Comment by bkhl** [comment]
I can now share my screen in Wayland, with:

Fedora Silverblue 33 (33.20210408.0)
gnome-shell-3.38.4-1
com.microsoft.Teams	1.4.00.7556 (from Flathub)

Unrelatedly, I also notice I'm not able to turn _off_ my camera with it, so be careful to use your camera flap if you have one …

**Comment by danshick** [comment]
That's interesting. There's nothing special going on in the flatpak file, so I believe the official Debian package must include support as well.

**Comment by richard-riverford** [comment]
I was all excited for a minute! I'm also running Teams 1.4.00.7556, from their RPM repo, but still _cannot_ share my screen. I'm running Fedora 34 Workstation with gnome-shell-40.0-1.

**Comment by danshick** [comment]
Hmmm, maybe I missed something in the flathub repo? I didn't see any clever flags or additional build steps or anything, but I may have missed something.

**Comment by danshick** [comment]
Looking at the commit history, this is the only screensharing cleverness I found...

https://github.com/flathub/com.microsoft.Teams/commit/d50457ed472a301808806898dd7fae0a19fc4e4c

Everything else for the past couple of months are just bot PRs with new versions of the `.deb` package.

**Comment by mario-tux** [comment]
I tested Teams 1.4.00.7556 using the AUR package (based on the official .deb package) and the flatpak one. No progress: in recent versions the sharing of windows/screen is not even proposed. 
I'm using Arch Linux with Gnome Shell 40 with Wayland.
Can someone else  confirm that this version of Teams works on Wayland in sharing the whole screen?

**Comment by billybraga** [comment]
I am using wayland on Ubuntu 20.04 and Gnome 3.36.7 and Teams 1.4.00.7556 **cannot** share my screen.

**Comment by dagbdagb** [comment]
Have a look at wf-recorder and v4l2loopback, as described over at [Arch](https://wiki.archlinux.org/index.php/Screen_capture#Via_a_virtual_webcam_video_feed). I have not tried myself yet. Looks promising.

**Comment by danshick** [comment]
I find the web-based Teams application is sufficient for screen sharing. The wf-recorder/v4l2loopback approach works okay too, but it is not as straight forward to set up. I was about to point out https://github.com/CatxFish/obs-v4l2sink, but it appears OBS now directly supports virtual camera mode. That's nice if you want to composite your webcam on top of your display as well (which wf-recorder can't do).

**Comment by vchernin** [comment]
[I've opened an upstream feature request to build Electron with PipeWire by default](https://github.com/electron/electron/issues/29115). Solving the upstream issue would theoretically solve this Teams issue, but does not necessarily mean PipeWire will work for Teams by default, currently you would still need `--enable-features=WebRTCPipeWireCapturer` (see the upstream issue for more info).

**Comment by hmsca** [comment]
Screen sharing is not working on Ubuntu 21.04. I'm using Wayland, MS Teams 1.4.00.7556 and Gnome 3.38.5.

**Comment by vchernin** [comment]
There doesn't appear to be anything we can do here to fix this issue. Since Microsoft Teams is proprietary we are dependent on them to fix this. Please upvote [this feature request](https://microsoftteams.uservoice.com/forums/555103-public/suggestions/41524504-screen-sharing-on-wayland) as likely this issue won't be fixed until Teams updates/fixes their version of Electron.

**Comment by ZurdenRo** [comment]
I use Fedora 34, Gnome 40 on Wayland. I cant screen shared, App dont crushes, but i cant see the option a button screen.

**Comment by igortobert** [comment]
Still can't share desktop

Manjaro 21.2, 
gnome 41.3, 
teams 1.4.00.26453-1,
wayland

**Comment by eiglow** [comment]
> There doesn't appear to be anything we can do here to fix this issue. Since Microsoft Teams is proprietary we are dependent on them to fix this. Please upvote [this feature request](https://microsoftteams.uservoice.com/forums/555103-public/suggestions/41524504-screen-sharing-on-wayland) as likely this issue won't be fixed until Teams updates/fixes their version of Electron.

Is it just me or is this link broken? If so, is there a new feature request we can follow?

**Comment by MightyCreak** [comment]
> Is it just me or is this link broken? If so, is there a new feature request we can follow?

They decommissioned their UserVoice instance... It's one way to fix all the user requests at once... I think the closest thing right now is the answers section in the MS docs: https://docs.microsoft.com/en-us/answers/search.html?type=question%20OR%20idea%20OR%20kbentry%20OR%20answer%20OR%20topic%20OR%20user&redirect=search%2Fsearch&sort=relevance&q=wayland%20%5Boffice-teams-linux-itpro%5D

**Comment by drew-parsons** [comment]
This Microsoft feedback link has more votes than the one in the previous comment, best to add your vote to it.

https://feedbackportal.microsoft.com/feedback/idea/4eacaa70-272e-ec11-b6e6-00224827bbc2

**Comment by drew-parsons** [comment]
Incidently for those wondering what Microsoft is doing with this (i.e. what version of electron they used), you can extract the version number from the installed files (in my case installed by the deb package)
```
grep -E "^electron" -A2 /usr/share/teams/resources/ThirdPartyNotice.txt
```
which currently indicates they're still using electron 10,
```
$ grep -E "^electron" -A2 /usr/share/teams/resources/ThirdPartyNotice.txt 
electron
10.4.7 <https://github.com/electron/electron/archive/v10.4.7.zip>
Copyright (c) 2013-2020 GitHub Inc.
```

**Comment by maxk096** [comment]
An update on this topic: As of December 2022, Microsoft deprecated MS Teams on Linux, favoring the PWA.

Sources:
- https://feedbackportal.microsoft.com/feedback/idea/0497b60d-6149-ed11-bba2-000d3a4e39a4
- https://techcommunity.microsoft.com/t5/microsoft-teams-blog/microsoft-teams-progressive-web-app-now-available-on-linux/ba-p/3669846?culture=en-us&country=us
